tokenizers/src/normalizers/precompiled.rs

```diff
@@ -0,0 +1,216 @@
+        write!(formatter, "Precompiled")
+    }
+
+    fn visit_map<A>(self, mut map: A) -> std::result::Result<Self::Value, A::Error>
+    where
+        A: serde::de::MapAccess<'de>,
+    {
+        let maybe_type = map.next_entry::<String, Vec<u8>>()?;
+        let maybe_type_str = maybe_type.as_ref().map(|(k, v)| (k.as_str(), v));
+        match maybe_type_str {
+            Some(("precompiled_charsmap", value)) => Ok(Precompiled::from(&value)
+                .map_err(|_| Error::custom("Cannot read `precompiled` string"))?),
+            _ => Err(Error::custom("Expected precompiled value, got {:?}")),
+        }
+    }
+}
+
+named!(be<u32>, u32!(Endianness::Little));
+
+pub type ArrayUnit = usize;
+
+trait ArrayUnitTrait {
+    fn has_leaf(&self) -> bool;
+    fn value(&self) -> isize;
+    fn label(&self) -> usize;
+    fn offset(&self) -> usize;
+}
+
+impl ArrayUnitTrait for ArrayUnit {
+    fn has_leaf(&self) -> bool {
+        (self >> 8) & 1 == 1
+    }
+
+    fn value(&self) -> isize {
+        (self & ((1usize << 31) - 1)) as isize
+    }
+
+    fn label(&self) -> usize {
+        self & ((1usize << 31) | 0xFF)
+    }
+
+    fn offset(&self) -> usize {
+        (self >> 10) << ((self & (1usize << 9)) >> 6)
+    }
+}
+
+type Array = Vec<ArrayUnit>;
+
+#[derive(Default, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct DoubleArray {
+    array: Array,
+}
+
+impl DoubleArray {
+    fn from(array: Array) -> Self {
+        Self { array }
+    }
+
+    pub fn common_prefix_search(&self, key: &[u8]) -> Vec<isize> {
+        let mut node_pos = 0;
+        let mut results = vec![];
+
+        let mut unit = self.array[node_pos];
+        node_pos ^= unit.offset();
+        for c in key {
+            if *c == 0u8 {
+                break;
+            }
+            node_pos ^= *c as usize;
+            unit = self.array[node_pos];
+            if unit.label() != *c as usize {
+                return results;
+            }
+            node_pos ^= unit.offset();
+            if unit.has_leaf() {
+                results.push(self.array[node_pos].value());
+            }
+        }
+        results
+    }
+}
+
+fn parse(precompiled_charsmap: &[u8]) -> IResult<&[u8], Array> {
+    let n = precompiled_charsmap.len();
+    let (mut rest, trie_size) = le_u32(precompiled_charsmap)?;
+    assert_eq!(trie_size % 4, 0);
+    // u8 to u32.
+    let trie_char_size = trie_size / 4;
+    let mut trie_blob = Vec::with_capacity(trie_char_size as usize);
+    for _ in 0..trie_char_size {
+        let (rest2, n) = le_u32(rest)?;
+        rest = rest2;
+        trie_blob.push(n.to_usize());
+    }
+    let normalized_blob = rest;
+    assert_eq!(rest.len() + trie_size as usize + 4, n);
+    Ok((normalized_blob, trie_blob))
+}
+
+impl Precompiled {
+    pub fn from(precompiled_charsmap: &[u8]) -> Result<Precompiled> {
+        let (normalized_blob, trie_blob) = parse(&precompiled_charsmap).unwrap();
```

Comment:
We should probably avoid the `.unwrap()` here too, no?

Reply:
Done